System/Posix/Files/Common.hsc

```diff
@@ -688,3 +718,287 @@ getFdPathVar (Fd fd) v =
+fileDaxX                   :: ExtendedFileStatus -> Bool
+
+#if HAVE_STATX
+testFlag :: ExtendedFileStatus -> CAttributes -> Bool
+testFlag ex flag =
+  let attributes = fileAttributesX ex
+      attributes_mask = fileAttributesMaskX ex
+  in (attributes .&. attributes_mask .&. flag) == flag
+
+fileCompressedX ex = testFlag ex (#const STATX_ATTR_COMPRESSED)
+fileImmutableX ex = testFlag ex (#const STATX_ATTR_IMMUTABLE)
+fileAppendX ex = testFlag ex (#const STATX_ATTR_APPEND)
+fileNoDumpX ex = testFlag ex (#const STATX_ATTR_NODUMP)
+fileEncryptedX ex = testFlag ex (#const STATX_ATTR_ENCRYPTED)
+#ifdef STATX_ATTR_VERITY
+fileVerityX ex = testFlag ex (#const STATX_ATTR_VERITY)
+#else
+{-# WARNING fileVerityX "fileVerityX: not available on this platform" #-}
+fileVerityX _ = error "fileVerityX not available on this platform"
+#endif
+#ifdef STATX_ATTR_DAX
+fileDaxX ex = testFlag ex (#const STATX_ATTR_DAX)
+#else
+{-# WARNING fileDaxX "fileDaxX: not available on this platform" #-}
+fileDaxX _ = error "fileDaxX not available on this platform"
+#endif
+
+fileNlinkX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_nlink)
+fileAttributesX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_attributes)
+fileBlockSizeX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_blksize)
+fileOwnerX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_uid)
+fileGroupX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_gid)
+fileModeX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_mode)
+fileIDX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_ino)
+fileSizeX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_size)
+fileBlocksX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_blocks)
+fileAttributesMaskX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ (#peek struct statx, stx_attributes_mask)
+fileAccessTimeHiResX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ \statx_ptr -> do
+    sec  <- (#peek struct statx, stx_atime.tv_sec) statx_ptr :: IO EpochTime
+    nsec <- (#peek struct statx, stx_atime.tv_nsec) statx_ptr :: IO (#type int)
+    let frac = toInteger nsec % 10^(9::Int)
+    return $ fromRational $ toRational sec + frac
+fileCreationTimeHiResX (ExtendedFileStatus statx) =
+  unsafePerformIO $ withForeignPtr statx $ \statx_ptr -> do
+    sec  <- (#peek struct statx, stx_btime.tv_sec) statx_ptr :: IO EpochTime
+    nsec <- (#peek struct statx, stx_btime.tv_nsec) statx_ptr :: IO (#type int)
+    let frac = toInteger nsec % 10^(9::Int)
+    return $ fromRational $ toRational sec + frac
```

Comment:
`Rational` is likely to be very slow. I'd suggest something like 
```
{-# LANGUAGE NumDecimals #-}
return $ secondsToNominalDiffTime $ MkFixed $ toInteger sec * 1e12 + toInteger nsec * 1e3
```
Could we also please add a test to check that time stamps are interpreted correctly? It's easy to swap `sec` and `nsec`, or treat them as a large `Int64` vs. two `Int32`.